### ISSUE TITLE
In ESLint 6, what is the logic behind for loops throwing "no-unused-vars" linter warning?

### ISSUE THREAD

**Issue by GasimGasimzada** [eslint/eslint#12121]
I am using CRA whose new version updates ESLint to v6. Now, I am getting errors in `for ... of` and `for ... in` loops. Here is an example:

```
if (graphQLErrors) {
  for (let err of graphQLErrors) { // this gives me error that "err" is unused
    if (err.extensions && err.extensions.status && err.extensions.status === 401) {
      return fromPromise(performRefresh()).flatMap(() => {
        return forward(operation);
      });
    }
  }
}
```

What's the logic and what does ESLint expect in this case?

**Comment by ooflorent** [comment]
I also encounter this warning using the following statements:

```js
const entrypoints = {};
for (const [chunk, entrypoint] of Object.entries(stats.entrypoints)) {
//          ^^^^^  ^^^^^^^^^^
//          flagged as unused
  const js = [];
  const css = [];
  for (const asset of entrypoint.assets) {
//           ^^^^^
//         also here
    if (asset.endsWith(".js")) {
      js.push(PUBLIC_PATH + asset);
    } else if (asset.endsWith(".css")) {
      css.push(PUBLIC_PATH + asset);
    }
  }
  entrypoints[chunk] = { js, css };
}
```

Works as expected in `6.1.0`
Broken in `6.2.0`

**Comment by g-plane** [comment]
Duplicate of #12117 . Please track that issue.
